aptos-move/framework/aptos-framework/sources/vesting.move

```diff
@@ -0,0 +1,1103 @@
+                VestEvent {
+                    admin: vesting_contract.admin,
+                    vesting_contract_address: contract_address,
+                    staking_pool_address: vesting_contract.staking.pool_address,
+                    period_vested,
+                    amount: vested_amount,
+                },
+            );
+        }
+    }
+
+    fun compute_vested_amount(vesting_contract: &VestingContract): u64 {
+        // Check if the next vested period has already passed.
+        let vesting_schedule = &vesting_contract.vesting_schedule;
+        let last_vested_period = vesting_schedule.last_vested_period;
+        let next_period_to_vest = last_vested_period + 1;
+        let last_completed_period =
+            (timestamp::now_seconds() - vesting_schedule.start_timestamp_secs) / vesting_schedule.period_duration;
+        // Short-circuit if the next period to vest has not ended yet.
+        if (last_completed_period < next_period_to_vest) {
+            return 0
+        };
+
+        // Calculate how much has vested, excluding rewards.
+        // Index is 0-based while period is 1-based so we need to subtract 1.
+        let schedule = &vesting_schedule.schedule;
+        let schedule_index = next_period_to_vest - 1;
+        let vesting_fraction = if (schedule_index < vector::length(schedule)) {
+            *vector::borrow(schedule, schedule_index)
+        } else {
+            // Last vesting schedule fraction will repeat until the grant runs out.
+            *vector::borrow(schedule, vector::length(schedule) - 1)
+        };
+        let total_grant = pool_u64::total_coins(&vesting_contract.grant_pool);
+        fixed_point32::multiply_u64(total_grant, vesting_fraction)
+    }
+
+    /// Distribute any withdrawable stake from the stake pool.
+    public entry fun distribute(contract_address: address) acquires VestingContract {
+        assert_active_vesting_contract(contract_address);
+
+        let vesting_contract = borrow_global_mut<VestingContract>(contract_address);
+        let coins = withdraw_stake(vesting_contract, contract_address);
+        let total_distribution_amount = coin::value(&coins);
+        if (total_distribution_amount == 0) {
+            coin::destroy_zero(coins);
+            return
+        };
+
+        // Distribute coins to all shareholders in the vesting contract.
+        let grant_pool = &vesting_contract.grant_pool;
+        let shareholders = &pool_u64::shareholders(grant_pool);
+        let len = vector::length(shareholders);
+        let i = 0;
+        while (i < len) {
+            let shareholder = *vector::borrow(shareholders, i);
+            let shares = pool_u64::shares(grant_pool, shareholder);
+            let amount = pool_u64::shares_to_amount_with_total_coins(grant_pool, shares, total_distribution_amount);
+            let share_of_coins = coin::extract(&mut coins, amount);
+            let recipient_address = get_beneficiary(vesting_contract, shareholder);
```

Comment:
hmm... this whole contract fails if this account is frozen or doesn't exist...

Reply:
set_beneficiary checks that the beneficiary account is registered for APT. I can add a nicer error if the account doesn't exist. As for frozen, that's a tricky problem. Admin can always change the beneficiary account though.